docs/yum.8
.\" yum - Yellowdog Updater Modified
.TH "yum" "8" ""  "Seth Vidal" ""
.SH "NAME"
yum \- Yellowdog Updater Modified
.SH "SYNOPSIS"
\fByum\fP [options] command [package ...]
.SH "DESCRIPTION"
.PP 
\fByum\fP is an interactive, rpm based, package manager. It can automatically
perform system updates, including dependency analysis and obsolete processing
based on "repository" metadata. It can also perform installation of new
packages, removal of old packages and perform queries on the installed and/or
available packages among many other commands/services (see below)\&. \fByum\fP
is similar to other high level package managers like apt\-get and smart\&.
.PP
While there are some graphical interfaces directly to the \fByum\fP code, more
recent graphical interface development is happening with PackageKit and the
gnome\-packagekit application\&.
.PP 
\fIcommand\fP is one of:
.br 
.I \fR * install package1 [package2] [\&.\&.\&.]
.br 
.I \fR * update [package1] [package2] [\&.\&.\&.]
.br 
.I \fR * update-to [package1] [package2] [\&.\&.\&.]
.br 
.I \fR * update-minimal [package1] [package2] [\&.\&.\&.]
.br 
.I \fR * check\-update
.br 
.I \fR * upgrade [package1] [package2] [\&.\&.\&.] 
.br
.I \fR * upgrade-to [package1] [package2] [\&.\&.\&.] 
.br
.I \fR * distribution-synchronization [package1] [package2] [\&.\&.\&.] 
.br
.I \fR * remove | erase package1 [package2] [\&.\&.\&.]
.br 
.I \fR * autoremove [package1] [\&.\&.\&.]
.br 
.I \fR * list [\&.\&.\&.]
.br 
.I \fR * info [\&.\&.\&.]
.br 
.I \fR * provides  | whatprovides feature1 [feature2] [\&.\&.\&.]
.br  
.I \fR * clean [ packages | metadata | expire-cache | rpmdb | plugins | all ]
.br
.I \fR * makecache [fast]
.br
.I \fR * groups  [\&.\&.\&.]
.br
.I \fR * search string1 [string2] [\&.\&.\&.]
.br
.I \fR * shell [filename]
.br
.I \fR * resolvedep dep1 [dep2] [\&.\&.\&.] 
    (maintained for legacy reasons only - use repoquery or yum provides)
.br
.I \fR * localinstall rpmfile1 [rpmfile2] [\&.\&.\&.] 
    (maintained for legacy reasons only - use install)
.br
.I \fR * localupdate rpmfile1 [rpmfile2] [\&.\&.\&.]
    (maintained for legacy reasons only - use update)
.br
.I \fR * reinstall package1 [package2] [\&.\&.\&.] 
.br
.I \fR * downgrade package1 [package2] [\&.\&.\&.] 
.br
.I \fR * deplist package1 [package2] [\&.\&.\&.] 
.br
.I \fR * repolist [all|enabled|disabled] 
.br
.I \fR * repoinfo [all|enabled|disabled] 
.br
.I \fR * repository-packages <enabled-repoid> <install|remove|remove-or-reinstall|remove-or-distribution-synchronization> [package2] [\&.\&.\&.]
.br
.I \fR * version [ all | installed | available | group-* | nogroups* | grouplist | groupinfo ]
.br
.I \fR * history [info|list|packages-list|packages-info|summary|addon-info|redo|undo|rollback|new|sync|stats] 
.br
.I \fR * load-transaction [txfile]
.br
.I \fR * updateinfo [summary | list | info | remove-pkgs-ts | exclude-updates | exclude-all | check-running-kernel]
.br
.I \fR * fssnapshot [summary | list | have-space | create | delete]
.br
.I \fR * fs [filters | refilter | refilter-cleanup | du]
.br
.I \fR * check
.br 
.I \fR * help [command] 
.br
.PP 
Unless the \-\-help or \-h option is given, one of the above commands
must be present\&.
.PP
Repository configuration is honored in all operations.
.PP 
.IP "\fBinstall\fP"
Is used to install the latest version of a package or
group of packages while ensuring that all dependencies are
satisfied\&.  (See \fBSpecifying package names\fP for more information) 
If no package matches the given package name(s), they are assumed to be a shell 
glob and any matches are then installed\&. If the name starts with @^ then it
is treated as an environment group (group install @^foo), an @ character and
it's treated as a group (plain group install)\&.

If the name starts with a "-" character, then a search is done within the
transaction and any matches are removed. Note that Yum options use the same
syntax and it may be necessary to use "--" to resolve any possible conflicts.

If the name is a file, then install works
like localinstall\&. If the name doesn't match a package, then package
"provides" are searched (e.g. "_sqlitecache.so()(64bit)") as are
filelists (Eg. "/usr/bin/yum"). Also note that for filelists, wildcards will
match multiple packages\&.

Because install does a lot of work to make it as easy as possible to use, there
are also a few specific install commands "\fBinstall-n\fP", "\fBinstall-na\fP"
and "\fBinstall-nevra\fP". These only work on package names, and do not process
wildcards etc.
.IP 
.IP "\fBupdate\fP"
If run without any packages, update will update every currently
installed package.  If one or more packages or package globs are specified, Yum will
only update the listed packages\&.  While updating packages, \fByum\fP
will ensure that all dependencies are satisfied\&. (See \fBSpecifying package names\fP for more information) 
If the packages or globs specified match to packages which are not currently installed then update will
not install them\&. update operates on groups, files, provides and filelists
just like the "install" command\&.

If the main obsoletes configure option is true (default) or the \-\-obsoletes
flag is present \fByum\fP will include package 
obsoletes in its calculations - this makes it better for distro\-version 
changes, for example: upgrading from somelinux 8.0 to somelinux 9.

Note that "\fBupdate\fP" works on installed packages first, and only if there
are no matches does it look for available packages. The difference is most
noticeable when you do "\fBupdate\fP foo-1-2" which will act exactly as
"\fBupdate\fP foo" if foo-1-2 is installed. You can use the "\fBupdate-to\fP"
if you'd prefer that nothing happen in the above case.
.IP 
.IP "\fBupdate-to\fP"
This command works like "\fBupdate\fP" but always specifies the version of the
package we want to update to.
.IP 
.IP "\fBupdate-minimal\fP"
This works like the update command, but if you have the package foo-1
installed and have foo-2 (bugfix) and foo-3 (enhancement) available with
updateinfo.xml then update-minimal --bugfix will update you to foo-2.
.IP 
.IP "\fBcheck\-update\fP"
Implemented so you could know if your machine had any updates that needed to
be applied without running it interactively. Returns exit value of 100 if
there are packages available for an update. Also returns a list of the packages
to be updated in list format. Returns 0 if no packages are available for
update. Returns 1 if an error occurred.
Running in verbose mode also shows obsoletes.
.IP
.IP "\fBupgrade\fP"
Is the same as the update command with the \-\-obsoletes flag set. See update 
for more details.
.IP 
.IP "\fBupgrade-to\fP"
This command works like "\fBupgrade\fP" but always specifies the version of the
package we want to update to.
.IP 
.IP "\fBdistribution\-synchronization\fP or \fBdistro\-sync\fP"
Synchronizes the installed package set with the latest packages available, this
is done by either obsoleting, upgrading or downgrading as appropriate. This will
"normally" do the same thing as the upgrade command however if you have the
package FOO installed at version 4, and the latest available is only
version 3, then this command will \fBdowngrade\fP FOO to version 3.

If you give the optional argument "full", then the command will also reinstall
packages where the install checksum and the available checksum do not match. And
remove old packages (can be used to sync. rpmdb versions). The optional argument
"different" can be used to specify the default operation.

This command does not perform operations on groups, local packages or negative
selections.
.IP 
.IP "\fBremove\fP or \fBerase\fP"
Are used to remove the specified packages from the system
as well as removing any packages which depend on the package being
removed\&. remove operates on groups, files, provides and filelists just like
the "install" command\&.(See \fBSpecifying package names\fP for more information) 

Note that "yum" is included in the protected_packages configuration, by default.
So you can't accidentally remove yum itself.

The remove_leaf_only configuration changes the behaviour of this command
to only remove packages which aren't required by something else.

The clean_requirements_on_remove configuration changes the behaviour of this
command to also remove packages that are only dependencies of this package.

Because remove does a lot of work to make it as easy as possible to use, there
are also a few specific remove commands "\fBremove-n\fP", "\fBremove-na\fP"
and "\fBremove-nevra\fP". These only work on package names, and do not process
wildcards etc.
.IP 
.IP "\fBautoremove\fP"
.IP 
With one or more arguments this command works like running the "\fBremove\fP"
command with the clean_requirements_on_remove turned on. However you can also
specify no arguments, at which point it tries to remove any packages that
weren't installed explicitly by the user and which aren't required by
anything (so called leaf packages).

Because autoremove does a lot of work to make it as easy as possible to use,
there are also a few specific autoremove commands "\fBautoremove-n\fP", 
"\fBautoremove-na\fP" and "\fBautoremove-nevra\fP". These only work on package
names, and do not process wildcards etc.
.IP "\fBlist\fP"
Is used to list various information about available
packages; more complete details are available in the \fIList Options\fP
section below\&.
.IP 
.IP "\fBprovides\fP or \fBwhatprovides\fP"
Is used to find out which package provides some feature
or file. Just use a specific name or a file-glob-syntax wildcards to list
the packages available or installed that provide that feature or file\&.
.IP 
.IP "\fBsearch\fP"
This is used to find packages when you know something about the package but
aren't sure of it's name. By default search will try searching just package
names and summaries, but if that "fails" it will then try descriptions and url.

Yum search orders the results so that those packages matching more terms will
appear first.

You can force searching everything by specifying "all" as the first argument.
.IP 
.IP "\fBinfo\fP"
Is used to list a description and summary information about available
packages; takes the same arguments as in the \fIList Options\fP
section below\&.
.IP 
.IP "\fBclean\fP"
Is used to clean up various things which accumulate in the
\fByum\fP cache directory over time.  More complete details can be found in
the \fIClean Options\fP section below\&.
.IP 
.IP "\fBmakecache\fP"
Is used to download and make usable all the metadata for the currently enabled
\fByum\fP repos. If the argument "fast" is passed, then we just try to make
sure the repos are current (much like "yum clean expire-cache").
.IP 
.IP "\fBgroups\fP"
A command, new in 3.4.2, that collects all the subcommands that act on groups
together. Note that recent yum using distributions (Fedora-19+, RHEL-7+) have
configured group_command=objects which changes how group commands act in some
important ways.

"\fBgroup install\fP" is used to install all of the individual packages in a
group, of the specified types (this works as if you'd taken each of those
package names and put them on the command line for a "yum install" command).
 The group_package_types configuration option specifies which types will
be installed.
 If you wish to "reinstall" a group so that you get a package that is currently
blacklisted the easiest way to do that currently is to install the package
manually and then run "groups mark packages-sync mygroup mypackagename" (or
use yumdb to set the group_member of the package(s)).

"\fBgroup update\fP" is just an alias for group install, when using
group_command=compat. This will install packages in the group not already
installed and upgrade existing packages. With group_command=simple it will just
upgrade already installed packages. With group_command=objects it will try to
upgrade the group object, installing any available packages not blacklisted
(marked '-' in group info) and will upgrade the installed packages.

"\fBgroup list\fP" is used to list the available groups from all \fByum\fP
repos. When group_command=objects the group is installed if the user
explicitly installed it (or used the group mark* commands to mark it installed).
It does not need to have any packages installed.
When not using group_command=objects groups are shown as "installed" if all
mandatory packages are installed, or if a group doesn't
have any mandatory packages then it is installed if any of the optional or
default package are installed (when not in group_command=objects mode).
You can pass optional arguments to the list/summary commands: installed,
available, environment, language, packages, hidden and ids (or any of those
prefixed by "no" to turn them off again).
Note that groups that are available but hidden will not be listed unless
\'hidden\' keyword is passed to the command.
If you pass the \-v option, to enable verbose mode, then the groupids are
displayed by default (but "yum group list ids" is often easier to read).

"\fBgroup remove\fP" is used to remove all of the packages in a group, unlike "groupinstall" this
will remove everything regardless of group_package_types. It is worth pointing
out that packages can be in more than one group, so "group install X Y" followed
by "group remove Y" does not do give you the same result as "group install X".

The groupremove_leaf_only configuration changes the behaviour of this command
to only remove packages which aren't required by something else.

"\fBgroup info\fP" is used to give the description and package list of a group (and which type
those packages are marked as). Note that you can use the yum-filter-data and
yum-list-data plugins to get/use the data the other way around (i.e. what
groups own packages need updating). If you pass the \-v option, to enable verbose
mode, then the package names are matched against installed/available packages
similar to the list command.

When using group_command=objects, the info command will display markers next
to each package saying how that package relates to the group object. The
meaning of these markers is:

.br
"-" = Package isn't installed, and won't be installed as part of the group (Eg.  "yum group install foo -pkgA" or "yum group install foo; yum remove pkgA" … this will have pkgA marked as '-')
.br
"+" = Package isn't installed, but will be the next time you run "yum upgrade" or "yum group upgrade foo"
.br
" " = Package is installed, but wasn't installed via the group (so "group remove foo" won't remove it).
.br
"=" = Package is installed, and was installed via the group.

you can move an installed package into an installed group using either
"group mark package-sync/package-sync-forced" or "yumdb set group_member".

"\fBgroup summary\fP" is used to give a quick summary of how many groups
are installed and available.

"\fBgroup mark\fP" and "\fBgroup unmark\fP" are used when groups are configured
in group_command=objects mode. These commands then allow you to alter yum's idea
of which groups are installed, and the packages that belong to them.

"\fBgroup mark install\fP" mark the group as installed. When
installed "\fByum upgrade\fP" and "\fByum group upgrade\fP" will install new
packages for the group (only those packages already installed will be marked as
members of the installed group to start with).

"\fBgroup mark remove\fP" the opposite of mark install.

"\fBgroup mark packages\fP" takes a group id (which must be installed) and marks
any given installed packages (which aren't members of a group) as members of
the group. Note that the data from the repositories does not need to specify
the packages as a member of the group.

"\fBgroup mark packages-force\fP" works like mark packages, but doesn't care if
the packages are already members of another group.

"\fBgroup mark blacklist\fP" will blacklist all packages marked to be installed
for a group. After this command a "yum group upgrade" will not install any new
packages as part of the group.

"\fBgroup mark convert-blacklist\fP"

"\fBgroup mark convert-whitelist\fP"

"\fBgroup mark convert\fP" converts the automatic data you get
without using groups as objects into groups as objects data, in other words
this will make "yum --setopt=group_command=objects groups list" look as similar
as possible to the current output of
"yum --setopt=group_command=simple groups list". This makes it much
easier to convert to groups as objects without having to reinstall. For groups
that are installed the whitelist variant will mark all uninstalled packages for
the group as to be installed on the next "yum group upgrade", the blacklist
variant (current default) will mark them all as blacklisted.

"\fBgroup unmark packages\fP" remove a package as a member from any groups.
.IP
.IP "\fBshell\fP"
Is used to enter the 'yum shell', when a filename is specified the contents of
that file is executed in yum shell mode. See \fIyum-shell(8)\fP for more info.
.IP
.IP "\fBresolvedep\fP"
Is used to list packages providing the specified dependencies, at most one
package is listed per dependency. This command is maintained for legacy
reasons only, use repoquery instead.
.IP
.IP "\fBlocalinstall\fP"
Is used to install a set of local rpm files. If required the enabled 
repositories will be used to resolve dependencies. Note that the install command

ChangeLog
2011-06-28  James Antill <james@and.org>

	* yum/__init__.py: Bump version to 3.4.3

2011-06-28  James Antill <james@and.org>

	* yum/__init__.py:  Don't call repo.doSetup() if we've already setup the repos. BZ
	 717163.   Test: yum in @core == has    setup.        yum rm @core
	 == hasn't setup.

2011-06-28  James Antill <james@and.org>

	* yum/history.py, yum/rpmtrans.py:  Disable the rpmtrans chroot workaround. Use exclusive locking in
	 history.   The only downside here appears to be that we get
	 .sqlite-journal files in /var/lib/yum/history ... which is probably
	 fine.  Also tested fine as a normal user, with read-only access.

2011-06-27  James Antill <james@and.org>

	* yummain.py: Change the logging for NFS path problems from debug to
	critical.

2011-06-24  Nick Jacek <njacek@redhat.com>

	* yummain.py: Checks if we have read and write access in the current
	directory. If not, chdir to /. BZ 698795

2011-06-23  Zdeněk Pavlas <zpavlas@redhat.com>

	* output.py: Make the source xgettext-friendly.  xgettext ignores
	_(var), so localize literals instead.

2011-06-23  James Antill <james@and.org>

	* cli.py: Don't show -c warning, when we've changed it due to
	--installroot. BZ 707358.

2011-06-22  James Antill <james@and.org>

	* yum/rpmtrans.py:  Workaround rpm chroot insanity, so we can use sqlite from
	 unInstStop() cb.   So the problem is that AGAIN, when we log stuff
	 for installed rpms we do it in _instCloseFile, which is outside the
	 chroot ... but when we log stuff for erase rpms we do it in
	 _unInstStop, which is inside the chroot.  sqlite will open it's
	 journal file, which is it has a working path for from outside the
	 chroot ... being inside the chroot breaks this, which results in
	 complete failure of everything.   Fix is to move out of the chroot,
	 before we call the history/sqlite APIs, and then move back in again
	 ... basically what rpm does.

2011-06-22  James Antill <james@and.org>

	* ChangeLog, yum.spec, yum/__init__.py: Update ChangeLog/version for
	3.4.2 release.

2011-06-22  Zdeněk Pavlas <zpavlas@redhat.com>

	* output.py: Clean up the 'Transaction summary' report.

2011-06-21  Seth Vidal <skvidal@fedoraproject.org>

	Merge branch 'yum-3_2_X' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum into yum-3_2_X  *
	'yum-3_2_X' of ssh://yum.baseurl.org/srv/projects/yum/git/yum:   Fix
	a leaked file descriptor

2011-06-21  Zdeněk Pavlas <zpavlas@redhat.com>

	* yum/__init__.py, yum/logginglevels.py: Fix a leaked file
	descriptor  'yum.log' is now closed when YumBase instance is GC'd.
	Resolves BZ 713924.

2011-06-21  Seth Vidal <skvidal@fedoraproject.org>

	* yum/packages.py: add check to make sure rest in the libc
	collapsing is not a single item list.

2011-06-20  Nick Jacek <njacek@redhat.com>

	* yum/__init__.py: Uses ngettext to pluralise messages produced in
	__init__.py. BZ 552317

2011-06-20  Nick Jacek <njacek@redhat.com>

	* cli.py: Uses ngettext to pluralise messages produced by several
	functions in cli.py. BZ 552317

2011-06-20  Nick Jacek <njacek@redhat.com>

	* output.py: Changes the Transaction Summary so that the plurals are
	printed using ngettext. BZ 552317

2011-06-20  James Antill <james@and.org>

	* yum/__init__.py: Add message to history redo/undo for when nothing
	is done.

2011-06-16  James Antill <james@and.org>

	* yum/rpmtrans.py: Turn a name into a txmbr, if we can. Helps erase
	filelog/scriptout and history.

2011-06-16  James Antill <james@and.org>

	* yum/__init__.py: Add comments on TS_FAILED setting.

2011-06-16  Nick Jacek <njacek@redhat.com>

	* yum/__init__.py, yum/rpmtrans.py: If verifyTransaction detects
	that an install or removal was not successful, it will now set the
	output_state of the pkg to TS_FAILED. BZ 661962  For the case of a
	removal, since the output_state will be TS_FAILED rather than just
	remaining TS_ERASE, yum will now correctly say that the removal
	failed instead of saying that the package was removed.

2011-06-16  James Antill <james@and.org>

	* yum/update_md.py: Add a __unicode__ to the UpdateNotice object, so
	the test code works.

2011-06-14  James Antill <james@and.org>

	* yum/__init__.py: Sort the packages, best we can easily, in
	searchPackageProvides().

2011-06-13  Nick Jacek <njacek@redhat.com>

	* yum/__init__.py, yum/depsolve.py:  Stops "Finished dependency resolution" from being printed multiple
	 times when --skip-broken is used. BZ 626569.

2011-06-13  Marcelo Moreira de Mello <mmello@redhat.com>

	* docs/yum.conf.5: - Fixed small typo at protected_multilib description on yum.conf man
	page  Hello,    This patches fixes a small type at
	protected_multilib description on yum.conf man page.   Cheers,
	Marcelo -- Marcelo Moreira de Mello  From: Marcelo Moreira de Mello
	<mmello@redhat.com> Date: Mon, 13 Jun 2011 16:10:07 -0300 Subject:
	[PATCH]  Fixed small typo at protected_multilib description at
	yum.conf man page

2011-06-13  James Antill <james@and.org>

	* yum/__init__.py: Show the correct "duplicate" package for search
	without --showduplicates.

2011-06-13  James Antill <james@and.org>

	* yum/packages.py: Always sort installed repos. after available
	repos.

2011-06-13  James Antill <james@and.org>

	* output.py: Show from_repo in matchcallback. Only print "Matched
	from" is there is more.

2011-06-08  Nick Jacek <njacek@redhat.com>

	* docs/yum.8: Updated documentation for the new groups command.

2011-06-08  Nick Jacek <njacek@redhat.com>

	* yumcommands.py: Remove the unused group command classes.

2011-06-08  Nick Jacek <njacek@redhat.com>

	* cli.py, yumcommands.py: Adds the Groups command, and removes
	individual command classes for grouplist etc. Adds the group summary
	function.

2011-06-06  Casey Jao <cjao@ugcs.caltech.edu>

	* test/complexremovetests.py: Add complex remove tests for new
	_remove_old_deps code.

2011-06-06  Casey Jao <cjao@ugcs.caltech.edu>

	* test/testbase.py: Make FakePackage use a simulated
	RPMDBAdditionalDataPackage for yumdb_info.


etc/yum.bash
# bash completion for yum

_yum_helper()
{
    local IFS=$'\n'
    if [[ -n "$YUM_CACHEDIR" && "$1 $2" == "list available" ]]; then
        for db in $(find "$YUM_CACHEDIR" -name primary_db.sqlite); do
            COMPREPLY+=( $( sqlite3 "$db" \
                "SELECT name||'.'||arch FROM packages WHERE name LIKE '$3%'"
            ) )
        done
        return
    fi
    COMPREPLY+=( $(
        /usr/share/yum-cli/completion-helper.py -d 0 -C "$@" 2>/dev/null ) )
}

_yum_list()
{
    # Fail fast for things that look like paths or options.
    [[ $2 == */* || $2 == [.~-]* ]] && return
    # Listing all available packages takes way too long
    [[ $1 != "installed" && ${#2} -lt 1 ]] && return
    _yum_helper list "$@"
}

# arguments:
#   1 = 1 or 0 to list enabled or disabled plugins
#   2 = current word to be completed
_yum_plugins()
{
    local val
    [[ $1 -eq 1 ]] && val='\(1\|yes\|true\|on\)' || val='\(0\|no\|false\|off\)'
    COMPREPLY+=( $( compgen -W '$( command grep -il "^\s*enabled\s*=\s*$val" \
        /etc/yum/pluginconf.d/*.conf 2>/dev/null \
        | sed -ne "s|^.*/\([^/]\{1,\}\)\.conf$|\1|p" )' -- "$2" ) )
}

# arguments:
#   1 = current word to be completed
_yum_binrpmfiles()
{
    local IFS=$'\n'
    COMPREPLY+=( $( compgen -f -o plusdirs -X '!*.rpm' -- "$1" ) )
    COMPREPLY=( $( compgen -W '"${COMPREPLY[@]}"' -X '*.src.rpm' ) )
    COMPREPLY=( $( compgen -W '"${COMPREPLY[@]}"' -X '*.nosrc.rpm' ) )
}

_yum_baseopts()
{
    local opts='--help --tolerant --cacheonly --config --randomwait
        --debuglevel --showduplicates --errorlevel --rpmverbosity --quiet
        --verbose --assumeyes --assumeno --version --installroot --enablerepo
        --disablerepo --exclude --disableexcludes --obsoletes --noplugins
        --nogpgcheck --skip-broken --color --releasever --setopt --downloadonly
        --downloaddir --disableincludes'
    [[ $COMP_LINE == *--noplugins* ]] || \
        opts+=" --disableplugin --enableplugin"
    printf %s "$opts"
}

_yum_transactions()
{
    COMPREPLY+=( $( compgen -W "$( $yum -d 0 -C history 2>/dev/null | \
        sed -ne 's/^[[:space:]]*\([0-9]\{1,\}\).*/\1/p' )" -- "$cur" ) )
}

_yum_atgroups()
{
    if [[ $1 == \@* ]]; then
        _yum_helper groups list all "${1:1}"
        COMPREPLY=( "${COMPREPLY[@]/#/@}" )
        return 0
    fi
    return 1
}

# arguments:
#   1 = current word to be completed
#   2 = previous word
# return 0 if no more completions should be sought, 1 otherwise
_yum_complete_baseopts()
{
    case $2 in

        -d|--debuglevel|-e|--errorlevel)
            COMPREPLY=( $( compgen -W '0 1 2 3 4 5 6 7 8 9 10' -- "$1" ) )
            return 0
            ;;

        --rpmverbosity)
            COMPREPLY=( $( compgen -W 'info critical emergency error warn
                debug' -- "$1" ) )
            return 0
            ;;

        -c|--config)
            local IFS=$'\n'
            COMPREPLY=( $( compgen -f -o plusdirs -X "!*.conf" -- "$1" ) )
            return 0
            ;;

        --installroot|--downloaddir)
            local IFS=$'\n'
            COMPREPLY=( $( compgen -d -- "$1" ) )
            return 0
            ;;

        --enablerepo)
            _yum_helper repolist disabled "$1"
            return 0
            ;;

        --disablerepo)
            _yum_helper repolist enabled "$1"
            return 0
            ;;

        --disableexcludes|--disableincludes)
            _yum_helper repolist all "$1"
            local main=
            [[ $2 == *excludes ]] && main=main
            COMPREPLY=( $( compgen -W '${COMPREPLY[@]} all $main' -- "$1" ) )
            return 0
            ;;

        --enableplugin)
            _yum_plugins 0 "$1"
            return 0
            ;;

        --disableplugin)
            _yum_plugins 1 "$1"
            return 0
            ;;

        --color)
            COMPREPLY=( $( compgen -W 'always auto never' -- "$1" ) )
            return 0
            ;;

        -R|--randomwait|-x|--exclude|-h|--help|--version|--releasever|--cve|\
        --bz|--advisory|--tmprepo|--verify-filenames|--setopt)
            return 0
            ;;

        --download-order)
            COMPREPLY=( $( compgen -W 'default smallestfirst largestfirst' \
                -- "$1" ) )
            return 0
            ;;

        --override-protection)
            _yum_list installed "$1"
            return 0
            ;;

        --verify-configuration-files)
            COMPREPLY=( $( compgen -W '1 0' -- "$1" ) )
            return 0
            ;;
    esac

    return 1
}

_yum()
{
    COMPREPLY=()
    local yum=$1 cur=$2 prev=$3 words=("${COMP_WORDS[@]}")
    declare -F _get_comp_words_by_ref &>/dev/null && \
        _get_comp_words_by_ref -n = cur prev words

    # Commands offered as completions
    local cmds=( autoremove check check-update clean deplist distro-sync
        downgrade erase fs groups help history info install list
        load-transaction makecache provides reinstall remove repolist search
        shell update upgrade version )

    local i c cmd subcmd
    for (( i=1; i < ${#words[@]}-1; i++ )) ; do
        [[ -n $cmd ]] && subcmd=${words[i]} && break
        # Recognize additional commands and aliases
        for c in ${cmds[@]} check-rpmdb distribution-synchronization erase \
            group groupinfo groupinstall grouplist groupremove groupupdate \
            grouperase install-na load-ts localinstall localupdate whatprovides ; do
            [[ ${words[i]} == $c ]] && cmd=$c && break
        done
    done

    case $cmd in

        autoremove|erase|remove)
            _yum_atgroups "$cur" || _yum_list installed "$cur"
            return 0
            ;;

        check|check-rpmdb)
            COMPREPLY=( $( compgen -W 'dependencies duplicates all' \
                -- "$cur" ) )
            return 0
            ;;

        check-update|makecache|resolvedep)
            return 0
            ;;

        clean)
            [[ $prev == $cmd ]] && \
                COMPREPLY=( $( compgen -W 'expire-cache packages headers
                    metadata cache dbcache all' -- "$cur" ) )
            return 0
            ;;

        deplist)
            local IFS=$'\n'
            COMPREPLY=( $( compgen -f -o plusdirs -X '!*.[rs]pm' -- "$cur" ) )
            _yum_list all "$cur"
            return 0
            ;;

        distro-sync|distribution-synchronization)
            [[ $prev == $cmd ]] && \
                COMPREPLY=( $( compgen -W 'full different' -- "$cur" ) )
            _yum_list installed "$cur"
            return 0
            ;;

        downgrade|reinstall)
            if ! _yum_atgroups "$cur" ; then
                _yum_binrpmfiles "$cur"
                _yum_list installed "$cur"
            fi
            return 0
            ;;

        fs)
            if [[ $prev == $cmd ]] ; then
                COMPREPLY=( $( compgen -W 'filters filter refilter
                    refilter-cleanup du status diff' -- "$cur" ) )
                return 0
            fi
            case $subcmd in
                filter)
                    COMPREPLY=( $( compgen -W 'languages documentation' \
                        -- "$cur" ) )
                    return 0
                    ;;
                refilter|refilter-cleanup)
                    _yum_list installed "$cur"
                    return 0
                    ;;
                du|status|diff)
                    local IFS=$'\n'
                    COMPREPLY=( $( compgen -f -o plusdirs -- "$cur" ) )
                    return 0
                    ;;
            esac
            return 0
            ;;

        group*)
            if [[ ($cmd == groups || $cmd == group) && $prev == $cmd ]] ; then
                COMPREPLY=( $( compgen -W 'info install list remove summary' \
                    -- "$cur" ) )
            else
                _yum_helper groups list all "$cur"
            fi
            return 0
            ;;

        help)
            [[ $prev == $cmd ]] && \
                COMPREPLY=( $( compgen -W '${cmds[@]}' -- "$cur" ) )
            return 0
            ;;

        history)
            if [[ $prev == $cmd ]] ; then
                COMPREPLY=( $( compgen -W 'info list packages-list
                    packages-info summary addon-info redo undo rollback new
                    sync stats' -- "$cur" ) )
                return 0
            fi
            case $subcmd in
                undo|repeat|addon|addon-info|rollback)
                    if [[ $prev == $subcmd ]]; then
                        COMPREPLY=( $( compgen -W "last" -- "$cur" ) )
                        _yum_transactions
                    fi
                    ;;
                redo)
                    case $prev in
                        redo)
                            COMPREPLY=( $( compgen -W "force-reinstall
                                force-remove last" -- "$cur" ) )
                            _yum_transactions
                            ;;
                        reinstall|force-reinstall|remove|force-remove)
                            COMPREPLY=( $( compgen -W "last" -- "$cur" ) )
                            _yum_transactions
                            ;;
                    esac
                    ;;
                package-list|pkg|pkgs|pkg-list|pkgs-list|package|packages|\
                packages-list|pkg-info|pkgs-info|package-info|packages-info)
                    _yum_list available "$cur"
                    ;;
                info|list|summary)
                    if [[ $subcmd != info ]] ; then
                        COMPREPLY=( $( compgen -W "all" -- "$cur" ) )
                        [[ $cur != all ]] && _yum_list available "$cur"
                    else
                        _yum_list available "$cur"
                    fi
                    _yum_transactions
                    ;;
                sync|synchronize)
                    _yum_list installed "$cur"
                    ;;
            esac
            return 0
            ;;

        info)
            _yum_list all "$cur"
            return 0
            ;;

        install)
            if ! _yum_atgroups "$cur" ; then
                _yum_binrpmfiles "$cur"
                _yum_list available "$cur"
            fi
            return 0
            ;;

        install-na)
            _yum_list available "$cur"
            return 0
            ;;

        list)
            [[ $prev == $cmd ]] && \
                COMPREPLY=( $( compgen -W 'all available updates installed
                    extras obsoletes recent' -- "$cur" ) )
            return 0
            ;;

        load-transaction|load-ts)
            local IFS=$'\n'
            COMPREPLY=( $( compgen -f -o plusdirs -X '!*.yumtx' -- "$cur" ) )
            return 0
            ;;

        localinstall|localupdate)
            _yum_binrpmfiles "$cur"
            return 0
            ;;

        provides|whatprovides)
            local IFS=$'\n'
            COMPREPLY=( $( compgen -f -o plusdirs -- "$cur" ) )
            return 0
            ;;

        repolist)
            [[ $prev == $cmd ]] && \
                COMPREPLY=( $( compgen -W 'all enabled disabled' -- "$cur" ) )
            return 0
            ;;

        search)
            [[ $prev == $cmd ]] && COMPREPLY=( $( compgen -W 'all' -- "$cur" ) )
            return 0
            ;;

        shell)
            if [[ $prev == $cmd ]]; then
                local IFS=$'\n'
                COMPREPLY=( $( compgen -f -o plusdirs -- "$cur" ) )
            fi
            return 0
            ;;

        update|upgrade)
            if ! _yum_atgroups "$cur" ; then
                _yum_binrpmfiles "$cur"
                _yum_list updates "$cur"
            fi
            return 0
            ;;
        version)
            [[ $prev == $cmd ]] && \
                COMPREPLY=( $( compgen -W 'all installed available nogroups
                    grouplist groupinfo' -- "$cur" ) )
            return 0
            ;;
    esac

    local split=false
    declare -F _split_longopt &>/dev/null && _split_longopt && split=true

    _yum_complete_baseopts "$cur" "$prev" && return 0

    $split && return 0

    if [[ $cur == -* ]] ; then
        COMPREPLY=( $( compgen -W '$( _yum_baseopts )' -- "$cur" ) )
        return 0
    fi
    COMPREPLY=( $( compgen -W '${cmds[@]}' -- "$cur" ) )
} &&
complete -F _yum -o filenames yum yummain.py

# Local variables:
# mode: shell-script
# sh-basic-offset: 4
# sh-indent-comment: t
# indent-tabs-mode: nil
# End:
# ex: ts=4 sw=4 et filetype=sh
